Source/BoundaryConditions/PML.cpp

```diff
@@ -660,6 +661,38 @@ PML::PML (const int lev, const BoxArray& grid_ba, const DistributionMapping& gri
         pml_G_fp->setVal(0.0);
     }
 
+    if (WarpX::em_solver_medium == MediumForEM::Macroscopic) {
+        // Number of guard cells is same as that required for E-field in pml and all macroscopic
+        // parameters are cell-centered.
+        pml_eps_fp = std::make_unique<amrex::MultiFab> (ba, dm, 1, nge);
+        pml_mu_fp = std::make_unique<amrex::MultiFab> (ba, dm, 1, nge);
+        pml_sigma_fp = std::make_unique<amrex::MultiFab> (ba, dm, 1, nge);
+
+        auto& warpx = WarpX::GetInstance();
+        auto& macroscopic_properties = warpx.m_macroscopic_properties;
+
+        if (macroscopic_properties->m_sigma_s == "constant") {
+            pml_sigma_fp->setVal(macroscopic_properties->m_sigma);
+        } else if (macroscopic_properties->m_sigma_s == "parse_sigma_function") {
+            macroscopic_properties->InitializeMacroMultiFabUsingParser(pml_sigma_fp.get(),
+                macroscopic_properties->m_sigma_parser->compile<3>(), lev);
+        }
+
+        if (macroscopic_properties->m_epsilon_s == "constant") {
+            pml_eps_fp->setVal(macroscopic_properties->m_epsilon);
+        } else if (macroscopic_properties->m_epsilon_s == "parse_epsilon_function") {
+            macroscopic_properties->InitializeMacroMultiFabUsingParser(pml_eps_fp.get(),
+                macroscopic_properties->m_epsilon_parser->compile<3>(), lev);
+        }
+
+        if (macroscopic_properties->m_mu_s == "constant") {
+            pml_mu_fp->setVal(macroscopic_properties->m_mu);
+        } else if (macroscopic_properties->m_mu_s == "parse_mu_function") {
+            macroscopic_properties->InitializeMacroMultiFabUsingParser(pml_mu_fp.get(),
+                macroscopic_properties->m_mu_parser->compile<3>(), lev);
+        }
+    }
+
```

Comment:
It seems to me that you never initialise the values of `pml_eps_cp`, `pml_mu_cp` and `pml_sigma_cp`, although they're used afterwards in the code. Is this done on purpose?

Reply:
Oh thank you so much for catching this!
I will add a separate PR (will add this in parallel so that it does make this one longer) to initialize the macro properties on the coarse patch.
The only reason this was not done is that we have never tested it with MR.
and actually, unless we have particles, the macroscopic solver may not work with mesh-refinement using the current WarpX algorithm. So I will add an abort, instead.

Action item:
1. Add an abort if macroscopic solver is selected with mesh-refinement
2. Either initialize macroscopic properties on coarse patch, or not use call the coarse patch for MR